llvm/lib/Transforms/Scalar/ConstraintElimination.cpp

```diff
@@ -534,20 +536,21 @@ static Decomposition decompose(Value *V,
     }
 
     if (match(V, m_NSWAdd(m_Value(Op0), m_Value(Op1))))
-      return MergeResults(Op0, Op1, IsSigned);
+      if (auto Decomp = MergeResults(Op0, Op1, IsSigned))
+        return *Decomp;
 
     if (match(V, m_NSWSub(m_Value(Op0), m_Value(Op1)))) {
       auto ResA = decompose(Op0, Preconditions, IsSigned, DL);
       auto ResB = decompose(Op1, Preconditions, IsSigned, DL);
-      ResA.sub(ResB);
-      return ResA;
+      if (!ResA.sub(ResB))
+        return ResA;
```

Comment:
Does it make sense to continue below after overflow? Should we just return std::nullopt?

Same for the others

Reply:
Done.
BTW, it is worth noting that `Preconditions` is not correctly maintained. Do you think it is valuable to fix this problem?

Reply:
Not correctly maintained as in contains preconditions that may not be needed as we faild with the decomposition?

Reply:
Yes. It may miss some optimizations. But it looks unlikely to happen in real-world cases.